web/source/text/engineDeviceSpec.ts

```diff
@@ -0,0 +1,76 @@
+namespace com.keyman.text {
+  export enum Browser {
+    Chrome = 'chrome',
+    Edge = 'edge',
+    Firefox = 'firefox',
+    Native = 'native', // Used by embedded mode
+    Opera = 'opera',
+    Safari = 'safari',
+    Other = 'other'
+  }
+
+  export enum OperatingSystem {
+    Windows = 'Windows',
+    macOS = 'MacOSX',
+    Linux = 'Linux',
+    Android = 'Android',
+    iOS = 'iOS',
+    Other = 'other'
+  }
+
+  export enum FormFactor {
+    Desktop = 'desktop',
+    Phone = 'phone',
+    Tablet = 'tablet'
+  }
+
+  /**
+   * This class provides an abstract version of com.keyman.Device that is headless-friendly, 
+   * containing only the information needed by web-core for text processing use.
+   */
+  export class EngineDeviceSpec {
+    readonly browser: Browser;
+    readonly formFactor: FormFactor;
+    readonly OS: OperatingSystem;
+    readonly touchable: boolean;
+
+    constructor(browser: string, formFactor: string, OS: string, touchable: boolean) {
+      switch(browser as Browser) {
+        case Browser.Chrome:
+        case Browser.Edge:
+        case Browser.Firefox:
+        case Browser.Native:
+        case Browser.Opera:
+        case Browser.Safari:
+          this.browser = browser as Browser;
+          break;
+        default:
+          this.browser = Browser.Other;
+      }
+
+      switch(formFactor.toLowerCase()) {
+        case FormFactor.Desktop:
+        case FormFactor.Phone:
+        case FormFactor.Tablet:
+          this.formFactor = formFactor as FormFactor;
```

Comment:
This won't, as far as I can tell, maintain the case as it should be set? So if I pass in `Desktop` then `this.formFactor` will be set to `Desktop` which will then not match against `FormFactor.Desktop` in other places.